# Project 4 — Cryptography (CS-GY 6903)

- Matthew Bentz (mb9661@nyu.edu)
- Darren Tahe (dt2607@nyu.edu)
- Srikanth Akella (ta2728@nyu.edu)
- Matthew Mobijohn (mm7655@nyu.edu)

---

## Part I. Merkle Trees and Data Integrity

### Task

Implement a binary Merkle tree using a cryptographic hash function such as SHA-256. Each leaf node of the tree should correspond to a plaintext file. The hash of each internal node is computed as the hash of the concatenation of its two children. 

### Experiments

1. Construct a Merkle tree using four leaf files and display the resulting tree structure and hashes. 

In [1]:
import os
import hashlib

# Helper functions

def merkle_hash(left, right):
    return hashlib.sha256((left + right).encode()).hexdigest()

def generate_files(path, amount):
    os.makedirs(path, exist_ok=True)
    for i in range(amount):
        with open(path + f"/file{i}.txt", "w") as f:
            f.write(f"This is file {i} at {path}.\n\nHere is a random string: {os.urandom(16).hex()}\n")

def get_file_hashes(path, amount):
    hashes = {}
    for i in range(amount):
        with open(path + f"/file{i}.txt", "rb") as f:
            data = f.read()
            hash_value = hashlib.sha256(data).hexdigest()
            hashes[i] = hash_value
    return hashes

In [2]:
path = "part1/four_leaf"
amount = 4
hashes = {}

# Generate 4 files for the Merkle tree

generate_files(path, amount)

# Hash each file

hashes = get_file_hashes(path, amount)

# Generate/display the Merkle tree

# Level 1 (leaves)
hash0 = hashes[0]
hash1 = hashes[1]
hash2 = hashes[2]
hash3 = hashes[3]
# Level 2
hash0_1 = merkle_hash(hash0, hash1)
hash2_3 = merkle_hash(hash2, hash3)
# Level 3 (root)
root_hash = merkle_hash(hash0_1, hash2_3)
print("Merkle Tree:")
print(f"\tRoot Hash: {root_hash}")
print(f"\t\tHash0_1: {hash0_1}")
print(f"\t\t\tHash0: {hash0}")
print(f"\t\t\tHash1: {hash1}")
print(f"\t\tHash2_3: {hash2_3}")
print(f"\t\t\tHash2: {hash2}")
print(f"\t\t\tHash3: {hash3}")

Merkle Tree:
	Root Hash: 8db758195661c0284466978d0f0afcd1cbff24afc47d6d62433f66358dadef4f
		Hash0_1: 79dcbae8a3da786a5291abf9d742e5e0b9bd60ae84a26f59e601e578e0b41877
			Hash0: 97cab24dd2216251cdf78ae964f8fb09e620264fc4568ef12b22832b0532d446
			Hash1: 717e4ede480c0df80231c510020c11268dd7a36a5a66c2cb33ca75b58ea0a9be
		Hash2_3: d02e9fac8206b2df812ca64e2664f1bf669329d05f7345713897441b62008d1c
			Hash2: 74996d7424308939c26b40a850a951b5f078f0be13407c15a1060e3f49d6f51c
			Hash3: 5f0fa6b1813f8136e32b9677d9a0bfcfcea843f054e711f386b0d882f8a51a4f


2. Construct a Merkle tree using six leaf files and display the tree structure and hashes.

In [3]:
path = "part1/six_leaf"
amount = 6
hashes = {}

# Generate 6 files for the Merkle tree

generate_files(path, amount)

# Hash each file

hashes = get_file_hashes(path, amount)

# Generate the Merkle tree

# Level 1 (leaves)
hash0 = hashes[0]
hash1 = hashes[1]
hash2 = hashes[2]
hash3 = hashes[3]
hash4 = hashes[4]
hash5 = hashes[5]
# Level 2
hash0_1 = merkle_hash(hash0, hash1)
hash2_3 = merkle_hash(hash2, hash3)
hash4_5 = merkle_hash(hash4, hash5)
# Level 3
hash0_1_2_3 = merkle_hash(hash0_1, hash2_3)
# Level 4 (root)
root_hash = merkle_hash(hash0_1_2_3, hash4_5)
print("Merkle Tree:")
print(f"\tRoot Hash: {root_hash}")
print(f"\t\tHash0_1_2_3: {hash0_1_2_3}")
print(f"\t\t\tHash0_1: {hash0_1}")
print(f"\t\t\t\tHash0: {hash0}")
print(f"\t\t\t\tHash1: {hash1}")
print(f"\t\t\tHash2_3: {hash2_3}")
print(f"\t\t\t\tHash2: {hash2}")
print(f"\t\t\t\tHash3: {hash3}")
print(f"\t\tHash4_5: {hash4_5}")
print(f"\t\t\tHash4: {hash4}")
print(f"\t\t\tHash5: {hash5}")

Merkle Tree:
	Root Hash: a29cfe8d9b8569f2456122c0929f5c6aeeb719218415ef7d1d5953e585b16819
		Hash0_1_2_3: 759e69ca4420baf1591d6e9348e067e579d182de4380670b53278d40486edd37
			Hash0_1: 2beff03cd3c3ae5cfb55896b4dbedd5bc17ef22178866148367a767e4ea14e0a
				Hash0: a334615f4e0a122d49f9a9862c17c5fe98944b076c6321363356e42b67b13d5c
				Hash1: 483ad8c693d3f3640c8ca5fa79755147b9c64774d86142d1f59b258ce2f9253f
			Hash2_3: 5c428ee297f108cbeddc309ea4f3012c3b6ef904740fb573312359f2d9b25d9f
				Hash2: 711aa942e8fde8f282ddb1c55f11ad86a17654141d8e82d9b241682db5b0e94a
				Hash3: 8b54839fe15c2a95b77cf79a4b9583304fa050869e6afffa3126fb1c215e12ca
		Hash4_5: 1835de75aa0248497507f405bed7730b194e3905882e142856e6a13a4e04f33c
			Hash4: 314611bcb20c069009ecd230729f712bcab3101110cd706ef831595e66a324e8
			Hash5: c4005f47596c0511f4ab595643ffe6250e1df593e119cb2877e9a4fee421c888


3. Modify the content of one of the files used in the four-leaf example. Recompute the tree and compare the new root hash with the original one.

In [4]:
path = "part1/four_leaf"
amount = 4
hashes = {}

# Generate/display the original Merkle tree

hashes = get_file_hashes(path, amount)

# Level 1 (leaves)
hash0 = hashes[0]
hash1 = hashes[1]
hash2 = hashes[2]
hash3 = hashes[3]
# Level 2
hash0_1 = merkle_hash(hash0, hash1)
hash2_3 = merkle_hash(hash2, hash3)
# Level 3 (root)
root_hash = merkle_hash(hash0_1, hash2_3)
print("Original Merkle Tree:")
print(f"\tRoot Hash: {root_hash}")
print(f"\t\tHash0_1: {hash0_1}")
print(f"\t\t\tHash0: {hash0}")
print(f"\t\t\tHash1: {hash1}")
print(f"\t\tHash2_3: {hash2_3}")
print(f"\t\t\tHash2: {hash2}")
print(f"\t\t\tHash3: {hash3}")


# manually modify the content of file0.txt and recompute the hash

with open(f"part1/four_leaf/file0.txt", "w") as f:
    f.write(f"This an update to the content in the file.\n\nHere is a random string: {os.urandom(16).hex()}\n")

# Generate/display the new Merkle tree

hashes = get_file_hashes(path, amount)

# Level 1 (leaves)
hash0 = hashes[0]
hash1 = hashes[1]
hash2 = hashes[2]
hash3 = hashes[3]
# Level 2
hash0_1 = merkle_hash(hash0, hash1)
hash2_3 = merkle_hash(hash2, hash3)
# Level 3 (root)
root_hash = merkle_hash(hash0_1, hash2_3)
print("Recomputed Merkle Tree:")
print(f"\tRoot Hash: {root_hash}")
print(f"\t\tHash0_1: {hash0_1}")
print(f"\t\t\tHash0: {hash0}")
print(f"\t\t\tHash1: {hash1}")
print(f"\t\tHash2_3: {hash2_3}")
print(f"\t\t\tHash2: {hash2}")
print(f"\t\t\tHash3: {hash3}")

Original Merkle Tree:
	Root Hash: 8db758195661c0284466978d0f0afcd1cbff24afc47d6d62433f66358dadef4f
		Hash0_1: 79dcbae8a3da786a5291abf9d742e5e0b9bd60ae84a26f59e601e578e0b41877
			Hash0: 97cab24dd2216251cdf78ae964f8fb09e620264fc4568ef12b22832b0532d446
			Hash1: 717e4ede480c0df80231c510020c11268dd7a36a5a66c2cb33ca75b58ea0a9be
		Hash2_3: d02e9fac8206b2df812ca64e2664f1bf669329d05f7345713897441b62008d1c
			Hash2: 74996d7424308939c26b40a850a951b5f078f0be13407c15a1060e3f49d6f51c
			Hash3: 5f0fa6b1813f8136e32b9677d9a0bfcfcea843f054e711f386b0d882f8a51a4f
Recomputed Merkle Tree:
	Root Hash: 2e90bf41821c935416395e856fb2a5d5a13175af520ced657886f439cf574b8a
		Hash0_1: f5297ea6ebd03d5b69c82683d942b48e94304bc22cb16be8e8ecce377e07a92a
			Hash0: 1b48c2a204632c2f315b9e50b01761edaa4cb3294d7cdacdee344e4d6c8449d1
			Hash1: 717e4ede480c0df80231c510020c11268dd7a36a5a66c2cb33ca75b58ea0a9be
		Hash2_3: d02e9fac8206b2df812ca64e2664f1bf669329d05f7345713897441b62008d1c
			Hash2: 74996d7424308939c26b40a850a951b5f078

### Questions

1. When the content of one leaf file is modified, which nodes in the tree change?

In this example, 3 nodes in the tree change. The hash of the original file that was modified, the has of file0/file1, and the root hash. In the above output, you can compare the Root Hash, Hash0_1, and Hash0.

2. Why does a small change in the file lead to a completely different root hash?

The small change in a file leads to a completely different root hash because of the avalanche effect properties in hashing functions, such as SHA-256 as used in the experiment. The drastic change in a leaf node will cascade recursively updwards, causing the root hash and any of it's derived children to the leaf to be significantly changed.

3. What properties of cryptographic hash functions make Merkle trees useful?

As described, the avalanche effect makes Merkle trees useful due to the drastic change in output. When a small change is reflected throughout the tree, you are able to traverse from the root to see exactly which file changed, but also have high confidence in the validity of the tree since the effect compounds. Additionally, collision resistance helps in this confidence, since its infeasible that a change will result in the same hash (and therefore, the same root).

4. Identify at least two real-world systems that make use of Merkle trees and briefly describe their purpose.

Two real-world systems that use Merkle trees would be blockchains (such as crypto-currencies, i.e. Bitcoin) and Git. In the Bitcoin example, each block is represented by a Merkle tree, and its purpose is to provide a lightweight and verifiable representation of the entire transaction list in the block. Additionally, the Merkle tree structure serves to allow $O(log_n)$ searches, rather than $O(n)$. For the Git example, Merkle trees allow commits to store the entire state of the repository by a hash (the trees root hash) at the specific point in time. Because we can detect at which level in the tree the commits changed, we can also efficiently identify the files that have changed in the repo between commits. 

---

## Part II. Hash Collisions and the Birthday Effect

### Task

Define a modified hash function based on SHA-256 where only the first k bits of the output are kept. Consider at least the following cases: k=4, k=8, k=12.

### Experiments

Generate random text inputs and compute their truncated hashes. Continue generating inputs until a collision is found.

For each value of k:

- record the number of attempts required to find a collision
- repeat the experiment several times if necessary

In [43]:
def find_collisions(k_bits: int, iterations: int) -> list:
    collision_counts = []
    for _ in range(iterations):
        attempts = 0
        hashes = set()
        while True:
            random_message = os.urandom(16).hex()
            full_hash = hashlib.sha256(random_message.encode()).digest()
            modified_hash = int.from_bytes(full_hash, byteorder='big') >> (256 - k_bits)
            attempts += 1
            if modified_hash in hashes:
                collision_counts.append(attempts)
                break
            hashes.add(modified_hash)
    return collision_counts

k4_iterations = find_collisions(4, 10000)
k8_iterations = find_collisions(8, 10000)
k12_iterations = find_collisions(12, 10000)

# k=4
print("k=4 Collision Statistics:")
print("Total output space size:", 2**4)
print("Expected number of attempts for a collision:", int(2**(4/2)))
print("Minimum attempts before collision:", min(k4_iterations))
print("Maximum attempts before collision:", max(k4_iterations))
print("Average attempts before collision:", sum(k4_iterations) / len(k4_iterations))
print()
# k=8
print("k=8 Collision Statistics:")
print("Total output space size:", 2**8)
print("Expected number of attempts for a collision:", int(2**(8/2)))
print("Minimum attempts before collision:", min(k8_iterations))
print("Maximum attempts before collision:", max(k8_iterations))
print("Average attempts before collision:", sum(k8_iterations) / len(k8_iterations))
print()
# k=12
print("k=12 Collision Statistics:")
print("Total output space size:", 2**12)
print("Expected number of attempts for a collision:", int(2**(12/2)))
print("Minimum attempts before collision:", min(k12_iterations))
print("Maximum attempts before collision:", max(k12_iterations))
print("Average attempts before collision:", sum(k12_iterations) / len(k12_iterations))
print()

k=4 Collision Statistics:
Total output space size: 16
Expected number of attempts for a collision: 4
Minimum attempts before collision: 2
Maximum attempts before collision: 15
Average attempts before collision: 5.6988

k=8 Collision Statistics:
Total output space size: 256
Expected number of attempts for a collision: 16
Minimum attempts before collision: 2
Maximum attempts before collision: 73
Average attempts before collision: 20.5951

k=12 Collision Statistics:
Total output space size: 4096
Expected number of attempts for a collision: 64
Minimum attempts before collision: 3
Maximum attempts before collision: 263
Average attempts before collision: 80.6503



### Questions

1. What is the minimum, maximum, and average number of attempts required to find a collision for each hash size?

Statistics defined in the python cell above.

2. How do your observations compare with the prediction from the birthday paradox?

Our results align well with the birthday paradox prediction. The birthday paradox tells us that collisions occur after roughly $2^{k/2}$ attempts.

From our experiment, we can see that for k bits of the hash kept, we expect to see:
- k=4:  $2^{4/2} = 2^2 = 4$ attempts
- k=8:  $2^{8/2} = 2^4 = 16$ attempts
- k=12: $2^{12/2} = 2^6 = 64$ attempts

The resulting averages are close to these approximations.

3. Explain why the expected number of attempts grows roughly on the order of $2^{(k/2)}$

The expected number of attempts grows roughly by $2^{(k/2)}$ because the possible pairs grows exponentially along with the total possible output space. Each new hash can collide with any of the previously computed hashes. A newly computed hash does not continue to fill the output space linearly as the output space grows, which results in the observation made by this experiment.

4. Why are collisions infeasible for modern hash functions such as SHA-256?

SHA-256 produces 256-bit hashes, meaning the output space has a size of $2^{256}$. Finding a collision still requires approximately $2^{256/2}$ or $2^{128}$ attempts. This is computationally infeasible and is the reason modern hash functions remain collision resistant.

---

## Part III. Hash Puzzles and Proof-of-Work

### Task

Construct a simple hash puzzle of the following form: compute hash(message || nonce) and search for a nonce such that the resulting hash begins with a specified number of zero bits.

### Experiments

Solve puzzles requiring:

1. one leading zero bit

2. two leading zero bits

3. three leading zero bits

Record the number of attempts required in each case.

In [41]:
def hash_puzzle(message: str, leading_zero_bits: int) -> tuple[int, str, int]:
    nonce = 0
    count = 0  # Counter to track the number of attempts
    while True:
        text = message + str(nonce)
        hash_result = hashlib.sha256(text.encode()).hexdigest()
        count += 1

        # technically, this "0" * leading_zero_bits check is comparing hex
        # so we are searching for 4 leading zero bits at a time
        if hash_result.startswith("0" * leading_zero_bits):
            return nonce, hash_result, count
        nonce += 1

one_leading_bit_attempts = []
two_leading_bit_attempts = []
three_leading_bit_attempts = []

for i in range(1000):
    nonce, hash_result, attempts = hash_puzzle(os.urandom(16).hex(), 1)
    one_leading_bit_attempts.append(attempts)
    
    nonce, hash_result, attempts = hash_puzzle(os.urandom(16).hex(), 2)
    two_leading_bit_attempts.append(attempts)

    nonce, hash_result, attempts = hash_puzzle(os.urandom(16).hex(), 3)
    three_leading_bit_attempts.append(attempts)

print("Average attempts to find 4 Leading Zero Bit: ", sum(one_leading_bit_attempts) / len(one_leading_bit_attempts))
print("Average attempts to find 8 Leading Zero Bits:", sum(two_leading_bit_attempts) / len(two_leading_bit_attempts))
print("Average attempts to find 12 Leading Zero Bits:", sum(three_leading_bit_attempts) / len(three_leading_bit_attempts))

_, _, attempts = hash_puzzle(os.urandom(16).hex(), 5)
print("Attempts to find 20 Leading Zero Bits:", attempts)

Average attempts to find 4 Leading Zero Bit:  16.412
Average attempts to find 8 Leading Zero Bits: 272.5
Average attempts to find 12 Leading Zero Bits: 3954.736
Attempts to find 20 Leading Zero Bits: 1018626


### Questions

1. How does the amount of work change as the number of required zero bits increases?

The amount of work required increases exponentially as the number of required leading zero bits increases. The expected number of attempts it takes to find k leading zero bits is roughly $2^k$. Puzzles with more leading zero bits will take longer to solve.

As you can see, the following sample output from the experiment follows this probability:

- Average attempts to find 4 Leading Zero Bit:  16.412
    - $2^4 = 16$
- Average attempts to find 8 Leading Zero Bits: 272.5
    - $2^8 = 256$
- Average attempts to find 12 Leading Zero Bits: 3954.736
    - $2^{12} = 4,096$
- Attempts to find 20 Leading Zero Bits: 1018626
    - $2^{20} = 1,048,576$

2. Why does the expected amount of work grow exponentially?

The work growing exponentially is caused by each additional leading zero bit reducing the previously probability of success by half. This is due to the fact that each bit has a 50% chance of being a 0 or 1. Therefore, for k required bits, our probability of success is roughly $(1/2)^k$.

3. Estimate the expected number of attempts needed for a puzzle requiring a 20-bit zero prefix.

As seen with our sample output, the expected number of attempts required for a 20-bit zero prefix is $2^{20} = 1,048,576$ attempts.

4. Explain how such puzzles are used in blockchain mining and why they help maintain security in decentralized systems.

Proof-of-work is when blockchain miners compete to solve hash puzzles by finding a nonce which produces a hash with a given leading number of zeros. They help maintain security because they are very difficult for attackers to manipulate. They also prevent spam or fraudulent transactions by ensuring that adding new blocks takes real computational effort. This keeps systems decentralized and secure by making attacks almost impossible without the correct computing power.

---

## Part IV. Reflection: Cryptography and AI

### Task

Ask an AI system to explain the following topics:

1. why hash collisions are inevitable

2. how Merkle trees work

3. how proof-of-work functions in blockain systems

Evaluate the response provided by the AI.

---

awaiting solution here

---

### Questions

1. Was the explanation correct and clear? Identify any mistakes or oversimplifications.

2. What aspects of the explanation were helpful?

3. In what ways might cryptographic techniques help improve trust in AI systems?

4. Conversely, how might AI tools assist in studying or analyzing crytpographic systems?